man/cw.conf.5.pod

  Copyright 2011-2015 James Hunt <james@jameshunt.us>

  This file is part of Clockwork.

  Clockwork is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  Clockwork is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with Clockwork.  If not, see <http://www.gnu.org/licenses/>.


=head1 NAME

cw.conf - Global Configuration for B<cw> Utility

=head1 SYNOPSIS

B</etc/clockwork/cw.conf>

=head1 DESCRIPTION

B<cw>(1) reads its configuration from I</etc/clockwork/cw.conf>, and then
attempts to read a similar file containing overrides, at I<~/.cwrc>.

Valid configuration directives, and their meanings, are as follows:

=over

=item B<mesh.master> - Mesh Master Server

Mesh master servers are the fixed points in the Clockwork Mesh architecture.
They are responsible for listening to clients, authenticating and relaying
queries to all subscribed B<cogd> nodes, giving back the appropriate
responses.

The master server must be identified by both its endpoint IP address and TCP
port number.  The standard Clockwork port for Mesh master servers is
TCP/2315.

=item B<mesh.cert> - Master Server Certificates

The Mesh master server will be authenticated by connecting clients via its
public certificate (which contains only the public key component).  This
helps to ensure that the B<cw> client is talking to the correct Mesh server.

=item B<mesh.timeout> - Mesh Response Timeout

How long (in seconds) B<cw> will continue to ask the Mesh master server for
replies to its query before it gives up and exits.  Since there is no way of
determining up-front how many nodes I<should> reply, B<cw> must wait out the
entire span of the timeout.

Defaults to I<40> seconds.  Values of 0 or less will be treated as this
default.

=item B<mesh.sleep> - Client Inter-check Sleep Interval

How long (in milliseconds) B<cw> will pause between result checks.  This
gives the Mesh server some respite from over-eager clients.

Defaults to I<250> milliseconds.  The minimum value is I<100> milliseconds.

=back

=head1 EXAMPLES

This is the default configuration, plus a single master server at
10.0.0.5:2314:

    mesh.timeout 40
    mesh.sleep   250

=head1 SEE ALSO

#SEEALSO

=head1 AUTHOR

#AUTHOR

=cut


CHANGELOG

origin/master                                              runtime 20150202

  (none)



3.0.10       2014-02-04                                    runtime 20150202

  [ENHANCEMENTS]

  - New Mesh commands for managing packages and services

    `package <name> <action> [version]` lets remote administrators manage
    and query the installed system packages.  Additionally, `package
    recache` will rebuild the package cache (i.e. apt-get update, or yum
    clean all).

    `service <name> <action>` deals with service run/boot status.



3.0.9        2014-02-02                                    runtime 20150202

  [ENHANCEMENTS]

  - New `sha1` opcode
    Useful for generating checksums of in-memory strings (like file names)

  - Modify `property` opcode to return pragma values
    If a property with the desired name isn't found, we should return the
    named pragma's value.  This lets us use pragmas as globals vars (without
    scope).  It also lets us move more functionality into stdlib.pn

  - Offline/Cache Mode
    cogd now supports cache/offline mode.  If a previous CFM run was
    successfull, the Pendulum bytecode image that represents the policy will
    be cached.  Subsequent runs that fail to connect to a clockd master can
    then fallback on this cached policy for local enforcement of the "last
    known good state"

  - Local File Caching support
    If files are marked 'cache: yes' in their resource definitions, cogd
    will keep the desired contents in a local cache directory, so that a
    future run can still enforce system state even when disconnected from a
    clockd policy master daemon.



3.0.8        2014-02-01                                    runtime 20150201

  [ENHANCMENTS]

  - New `augeas.exists?` opcode
    Like the 20150119 opcode `augeas.find`, except that it has no output
    parameter and therefore does no heap-allocation.

  - Introduce proper JSON and XML/CDATA escaping in mesh output routines



3.0.7        2014-02-01                                    runtime 20150131

  [ENHANCEMENTS]

  - Set a umask in cogd
    cogd now sets a more sane umask, which can be specified via a new
    configuration directive, 'umask'.  Defaults to 0022.

  - pn --output flag
    The pn utility, in assembler (-S) mode, now supports -o / --output
    for specifying where to write the pendulum bytecode image.  This allows
    callers to short-circuit the auto-naming ("just add .S") logic, and was
    needed to support VPATH builds in automake.

  - New clockwork.* facts
    The new clockwork.version, clockwork.runtime and clockwork.protocol
    facts can be used to query for version (x.y.z), runtime (YYYYMMDD) and
    network protocol version (N) in play.

  - New `runtime` opcode for determining version of Pendulum runtime
    The runtime version will be set to date of finalization of new
    instruction set prior to release.  This means the each runtime version
    is a fixed specification of interface (but not necessarily behavior),
    and pendulum code can use this new opcode to determine if the runtime is
    new enough before using new opcodes.

  - New `fs.mkparent` opcode
    Intelligently creates missing parent directories of its first argument,
    using the ownership and mode of the closest existing ancestor directory.
    This emulates pre-3.0 Clockwork behavior that was missed during the
    conversion to the register machine implementation.

  - Better error / output handling from `localsys` in stdlib.pn

  [BUG FIXES]

  - Memory leaks plugged
    cogd no longer leaks the bytecode image it received from clockd (big).
    SHA1.FAIL PDUs are no longer leaked by `remote.sha1` opcode.
    SHA1 checksums are no longer leaked by `remote.sha1` opcode.

  - Fix some vim pnasm highlighting issues

  - Remove superfluous debugging from stdlib.pn



3.0.6        2014-01-30                                    runtime 20150119

  [BUG FIXES]

  - Fix misfiring notifies for dependency forward tracking.
    Now, if a service depends on a config file, changing the config file
    will cause cogd to properly reload/restart the dependent service.

  - Properly set "changed" (lastchg) attribute on new users
    Without it, new users are required to change their passwords immediately
    upon login.



3.0.5        2014-01-29                                    runtime 20150119

  [BUG FIXES]

  - Fix an issue in stdlib.pn's res.file.contents function, whereby file
    resources that come with verification steps (as /etc/sudoers often do)
    don't transfer mode or ownership (UID/GID) to the temporary file before
    renaming it over the top of the original file.  This causes havoc on
    files like /etc/sudoers, which *must* have specific permissions.

  - Fix typo (retrieve not 'retreive') in diagnostic message for remote.file.



3.0.4        2014-01-28                                    runtime 20150119

  [BUG FIXES]

  - cogd now honors `timeout` when waiting for responses from clockd
    masters, during remediation.  This fixes several stall-out conditions
    where the clockd master goes away mid-run, and the cogd process hangs
    waiting for a response that will never come.



3.0.3        2014-01-27                                    runtime 20150119

  [ENHANCEMENTS]

  - SIGHUP-to-reload Support for cogd
    cogd supports use of `kill -HUP $pid` to cause it to reload
    configuration while running.

  - SIGHUP-to-reload Support for clockd
    clockd supports use of `kill -HUP $pid` to cause it to reload
    configuration and parse a new policy manifest while running.
    Intelligently handles client connections for in-flight remediation.



3.0.2        2014-01-22                                    runtime 20150119

  [ENHANCEMENTS]

  - Support '-%>' for newline suppression in ERB templates
    Any ERB ruby snippet ending with '-%>' (instead of just '%>'), will
    ignore trailing whitespace.

  [BUG FIXES]


src/cw.c
/*
  Copyright 2011-2015 James Hunt <james@jameshunt.us>

  This file is part of Clockwork.

  Clockwork is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  Clockwork is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with Clockwork.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "../src/clockwork.h"
#include "../src/policy.h"
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <getopt.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <uuid/uuid.h>

#ifdef CW_CLI_DIR
#  define DEFAULT_EXEC_PATH  CW_CLI_DIR
#else
#  define DEFAULT_EXEC_PATH  "/lib/clockwork/cli"
#endif

static void usage(int rc)
{
	fprintf(stderr, "Usage: cw [--version] [--help|-h]\n"
	                "          [--exec <path>] [--dump]\n"
	                "          <command> [args]\n"
	                "\n"
	                "Commonly used commands:\n"
	                "  fact      Show local system facts\n"
	                "  cert      Generate certificates (for host identity)\n"
	                "  shell     Interactively explore a manifest\n"
	                "  trust     Manage a certificate trust database\n"
	                "\n"
	                "Mesh commands (for interacting with a network of machines):\n"
	                "\n"
	                "  show      Show version and ACL information\n"
	                "  ping      Gauge host response time\n"
	                "  query     Gather info about node-local resources\n"
	                "  exec      Execute arbitrary commands\n"
	                "  package   Manage package installations\n"
	                "  service   Start and stop services\n"
	                "  cfm       Trigger configuration management runs\n"
	                "\n"
	                "'cw help -c' lists available commands. and 'cw help -g' lists\n"
	                "some helpful Clockwork guides.  Run 'cw help <command>' and\n"
	                "'cw help <guide> to read the specifics.\n"
	                "\n");

	if (rc >= 0)
		exit(rc);
}

static int get_base_options(int argc, char **argv)
{
	char *exec_path = strdup(DEFAULT_EXEC_PATH);

	const char *short_opts = "+h";
	struct option long_opts[] = {
		{ "help",    no_argument,       NULL, 'h' },
		{ "version", no_argument,       NULL,  1  },
		{ "exec",    required_argument, NULL,  2  },
		{ "dump",    no_argument,       NULL,  3  },
		{ 0, 0, 0, 0 },
	};

	int opt, idx = 0;
	while ((opt = getopt_long(argc, argv, short_opts, long_opts, &idx)) != -1) {
		switch (opt) {
		case 'h':
		case '?':
			usage(0);

		case 1: /* --version */
			printf("Clockwork version %s\n", PACKAGE_VERSION);
			exit(0);

		case 2: /* --exec */
			free(exec_path);
			exec_path = strdup(optarg);
			break;

		case 3: /* --dump */
			printf("exec_path %s\n", DEFAULT_EXEC_PATH);
			exit(0);

		default:
			usage(1);
		}
	}

	char *old_path = getenv("PATH");
	char *path = old_path ? string("%s:%s", exec_path, old_path)
	                      : strdup(exec_path);
	setenv("PATH", path, 1);
	free(exec_path);
	free(path);

	if (optind >= argc)
		usage(0);

	return optind;
}

static int builtin_cw_cert    (int argc, char **argv);
static int builtin_cw_cfm     (int argc, char **argv);
static int builtin_cw_fact    (int argc, char **argv);
static int builtin_cw_help    (int argc, char **argv);
static int builtin_cw_trust   (int argc, char **argv);
static int builtin_cw_uuid    (int argc, char **argv);

static struct {
	const char *cmd;
	int (*fn)(int,char**);
} BUILTINS[] = {
	{ "cw-cert",      builtin_cw_cert      },
	{ "cw-cfm",       builtin_cw_cfm       },
	{ "cw-fact",      builtin_cw_fact      },
	{ "cw-help",      builtin_cw_help      },
	{ "cw-trust",     builtin_cw_trust     },
	{ "cw-uuid",      builtin_cw_uuid      },
	{ NULL, NULL },
};

int main(int argc, char **argv)
{
	char *bin = strrchr(argv[0], '/');
	if (bin) bin++;
	else bin = argv[0];

	int i;
	for (i = 0; BUILTINS[i].cmd; i++)
		if (strcmp(bin, BUILTINS[i].cmd) == 0)
			return (*BUILTINS[i].fn)(argc, argv);

	if (strncmp(bin, "cw-", 3) == 0) {
		fprintf(stderr, "cw: '%s' is not a Clockwork builtin.  See 'cw --help'.\n", bin);
		return 2;
	}

	/* dispatch via exec! */
	int next = get_base_options(argc, argv);
	argv[next] = string("cw-%s", argv[next]);
	execvp(argv[next], argv+next);

	fprintf(stderr, "cw: '%s' is not a Clockwork command.  See 'cw --help'.\n", argv[next]);
	return 1;
}

static int builtin_cw_cert(int argc, char **argv)
{
	const char *short_opts = "h?f:i:u";
	struct option long_opts[] = {
		{ "help",     no_argument,       NULL, 'h' },
		{ "identity", required_argument, NULL, 'i' },
		{ "file",     required_argument, NULL, 'f' },
		{ "user",     no_argument,       NULL, 'u' },
		{ 0, 0, 0, 0 },
	};

	char *ident = NULL;
	char *file  = strdup("cwcert");
	int   type  = VIGOR_CERT_ENCRYPTION;

	int opt, idx = 0;
	while ( (opt = getopt_long(argc, argv, short_opts, long_opts, &idx)) != -1) {
		switch (opt) {
		case 'h':
		case '?':
			printf("USAGE: %s [--identity FQDN] [--type host|user] [--file cwcert]\n", argv[0]);
			exit(0);

		case 'i':
			free(ident);
			ident = strdup(optarg);
			break;

		case 'f':
			free(file);
			file = strdup(optarg);
			break;

		case 'u':
			type = VIGOR_CERT_SIGNING;
			break;
		}
	}

	char *pubfile = string("%s.pub", file);
	int pubfd = open(pubfile, O_WRONLY|O_CREAT|O_EXCL, 0444);
	if (pubfd < 0) {
		perror(pubfile);
		exit(1);
	}
	int secfd = open(file, O_WRONLY|O_CREAT|O_EXCL, 0400);
	if (secfd < 0) {
		perror(file);
		unlink(pubfile);
		exit(1);
	}

	FILE *pubio = fdopen(pubfd, "w");
	FILE *secio = fdopen(secfd, "w");
	if (!pubio || !secio) {
		perror("fdopen");
		fclose(pubio); close(pubfd); unlink(pubfile);
		fclose(secio); close(secfd); unlink(file);
		exit(1);
	}

	cert_t *cert = cert_generate(type);
	assert(cert);
	if (ident) {
		cert->ident = ident;
	} else if (cert->type == VIGOR_CERT_SIGNING) {
		cert->ident = getenv("USER") ? strdup(getenv("USER")) : NULL;
	} else {
		cert->ident = fqdn();
	}
	if (!cert->ident || strlen(cert->ident) == 0) {
		fprintf(stderr, "Failed to determine certificate/key identity!\n");
		fclose(pubio); close(pubfd); unlink(pubfile);
		fclose(secio); close(secfd); unlink(file);
		exit(1);
	}

	cert_writeio(cert, pubio, 0);
	cert_writeio(cert, secio, 1);

	fclose(pubio);
	fclose(secio);

	cert_free(cert);
	return 0;
}

static int builtin_cw_cfm(int argc, char **argv)
{
	return 0;
}

static int builtin_cw_fact(int argc, char **argv)
{
	char *config_file = NULL;
	int brief = 0;

	const char *short_opts = "h?vqVc:b";
	struct option long_opts[] = {
		{ "help",        no_argument,       NULL, 'h' },
		{ "config",      required_argument, NULL, 'c' },
		{ "brief",       no_argument,       NULL, 'b' },
		{ 0, 0, 0, 0 },
	};
	int opt, idx = 0;
	while ( (opt = getopt_long(argc, argv, short_opts, long_opts, &idx)) != -1) {
		switch (opt) {
		case 'h':
		case '?':
			printf("usage: cw-fact [--config <path>] [--brief|-b] [fact(s)]\n");
			exit(0);

		case 'c':
			free(config_file);
			config_file = strdup(optarg);
			break;

		case 'b':
			brief = 1;
			break;
		}
	}

	LIST(config);
	config_set(&config, "gatherers", CW_GATHER_DIR "/*");
	FILE *io = NULL;

	if (config_file) {
		io = fopen(config_file, "r");
		if (!io) {
			fprintf(stderr, "%s: %s\n", config_file, strerror(errno));
			exit(2);
		}
	} else {
		io = fopen(CW_COGD_CONFIG_FILE, "r");
	}

	if (io) {
		if (config_read(&config, io) != 0) {
			fprintf(stderr, "%s: failed to parse\n", config_file);
			exit(3);
		}
		fclose(io);
	}

	char *gatherers = config_get(&config, "gatherers");

	hash_t facts;
	memset(&facts, 0, sizeof(hash_t));
	if (fact_gather(gatherers, &facts) != 0) {
		fprintf(stderr, "Failed to gather facts via %s\n", gatherers);
		exit(4);
	}

	int rc = 0;
	if (optind >= argc) {
		fact_write(stdout, &facts);

	} else {
		char *v;
		for (; optind < argc; optind++) {
			v = hash_get(&facts, argv[optind]);
			if (v) {
				if (brief)
					printf("%s\n", v);
				else
					printf("%s=%s\n", argv[optind], v);

			} else {
				fprintf(stderr, "No such fact '%s'\n", argv[optind]);
				rc = 5;
			}
		}
	}

	return rc;
}

static int builtin_cw_help(int argc, char **argv)
{
	return 0;
}

static int builtin_cw_trust(int argc, char **argv)
{
	const char *short_opts = "h?trd:";
	struct option long_opts[] = {
		{ "help",     no_argument,        NULL, 'h' },
		{ "trust",    no_argument,        NULL, 't' },
		{ "revoke",   no_argument,        NULL, 'r' },
		{ "database", required_argument,  NULL, 'd' },
		{ 0, 0, 0, 0 },
	};

	char mode = 't';
	char *path = strdup("/etc/clockwork/certs/trusted");

	int opt, idx = 0;
	while ( (opt = getopt_long(argc, argv, short_opts, long_opts, &idx)) != -1) {
		switch (opt) {
		case 'h':
		case '?':
			printf("USAGE: %s [OPTIONS] FILE1 FILE2\n", argv[0]);
			exit(0);

		case 't':
		case 'r':
			mode = opt;
			break;

		case 'd':
			free(path);
			path = strdup(optarg);
			break;
		}
	}

	trustdb_t *db = trustdb_read(path);
	if (!db) db = trustdb_new();

	int rc;
	int i, n;
	for (i = optind, n = 0; i < argc; i++) {
		cert_t *cert = cert_read(argv[i]);
		if (!cert) {
			fprintf(stderr, "skipping %s: %s\n", argv[i], strerror(errno));
			continue;
		}

		if (!cert->pubkey) {
			fprintf(stderr, "skipping %s: no public key found in certificate\n", argv[i]);
			cert_free(cert);
			continue;
		}

		if (mode == 't') {
			rc = trustdb_trust(db, cert);
			assert(rc == 0);
			printf("TRUST %s %s\n", cert->pubkey_b16, cert->ident ? cert->ident : "(no ident)");
		} else {
			rc = trustdb_revoke(db, cert);
			assert(rc == 0);
			printf("REVOKE %s %s\n", cert->pubkey_b16, cert->ident ? cert->ident : "(no ident)");
		}
		cert_free(cert);
		n++;
	}

	printf("Processed %i certificate%s\n", n, n == 1 ? "" : "s");
	if (!n) exit(0);

	rc = trustdb_write(db, path);
	if (rc != 0) {
		fprintf(stderr, "Failed to write out trustdb %s: %s\n", path, strerror(errno));
		exit(2);
	}

	printf("Wrote %s\n", path);
	free(path);
	trustdb_free(db);
	return 0;
}

#define UUID_FILE  CW_SYSCONF_DIR "/.uuid"

static int builtin_cw_uuid(int argc, char **argv)
{
	const char *short_opts = "h?";
	struct option long_opts[] = {
		{ "help",     no_argument,        NULL, 'h' },
		{ "regen",    no_argument,        NULL,  1 },
		{ 0, 0, 0, 0 },
	};

	int regen = 0;

	int opt, idx = 0;
	while ( (opt = getopt_long(argc, argv, short_opts, long_opts, &idx)) != -1) {
		switch (opt) {
		case 'h':
		case '?':
			printf("USAGE: %s [--regen]\n", argv[0]);
			exit(0);

		case 1:
			regen = 1;
			break;
		}
	}

	char s[37];
	FILE *io;
	uuid_t uuid;

	if (regen) {
regen:
		io = fopen(UUID_FILE, "w");
		if (!io) {
			perror(UUID_FILE);
			exit(1);
		}

		uuid_generate(uuid);
		uuid_unparse_lower(uuid, s);
		fprintf(io, "%s\n", s);
		fclose(io);

	} else {
		io = fopen(UUID_FILE, "r");
		if (!io) {
			if (errno == ENOENT) goto regen;
			perror(UUID_FILE);
			exit(2);
		}

		if (!fgets(s, 37, io)) {
			perror(UUID_FILE);
			exit(3);
		}
		s[36] = '\0';
		fclose(io);
	}

	printf("%s\n", s);
	return 0;
}
